File: core/src/com/symbol/game/map/camera/Background.java
package com.symbol.game.map.camera;

import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.symbol.game.Config;

public class Background {

    private TextureRegion bgTexture;
    private OrthographicCamera cam;
    private Vector2 scale;

    private Vector2 position = new Vector2();
    private Vector2 velocity = new Vector2();

    private int numDrawX;
    private int numDrawY;

    public Background(TextureRegion bgTexture, OrthographicCamera cam, Vector2 scale) {
        this.bgTexture = bgTexture;
        this.cam = cam;
        this.scale = scale;

        numDrawX = (Config.V_WIDTH * 2) / bgTexture.getRegionWidth() + 1;
        numDrawY = (Config.V_HEIGHT * 2) / bgTexture.getRegionHeight() + 1;
        fixBleeding(bgTexture);
    }

    public void update(float dt) {
        position.x += (velocity.x * scale.x) * dt;
        position.y += (velocity.y * scale.y) * dt;
    }

    public void render(Batch batch) {
        float x = ((position.x + cam.viewportWidth / 2 - cam.position.x) * scale.x) % bgTexture.getRegionWidth();
        float y = ((position.y + cam.viewportHeight / 2 - cam.position.y) * scale.y) % bgTexture.getRegionHeight();

        int colOffset = x > 0 ? -1 : 0;
        int rowOffset = y > 0 ? -1 : 0;

        for (int row = 0; row < numDrawY; row++) {
            for (int col = 0; col < numDrawX; col++) {
                batch.draw(bgTexture,
                        x + (col + colOffset) * bgTexture.getRegionWidth(),
                        y + (row + rowOffset) * bgTexture.getRegionHeight());
            }
        }
    }

    private void fixBleeding(TextureRegion texture) {
        float fix = 0.01f;
        float invWidth = 1f / texture.getTexture().getWidth();
        float invHeight = 1f / texture.getTexture().getHeight();
        texture.setRegion((texture.getRegionX() + fix) * invWidth,
                (texture.getRegionY() + fix) * invHeight,
                (texture.getRegionX() + texture.getRegionWidth() - fix) * invWidth,
                (texture.getRegionY() + texture.getRegionHeight() - fix) * invHeight);
    }

}

File: core/src/com/symbol/map/Background.kt
package com.symbol.map

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Vector2
import com.symbol.game.Config

class Background(private val bgTexture: TextureRegion, private val cam: OrthographicCamera, private val scale: Vector2) {

    private var position: Vector2 = Vector2()
    private var velocity: Vector2 = Vector2()

    private var numDrawX: Int = 0
    private var numDrawY: Int = 0

    init {
        numDrawX = Config.V_WIDTH / bgTexture.regionWidth + 1
        numDrawY = Config.V_HEIGHT / bgTexture.regionHeight + 1
        fixBleeding(bgTexture)
    }

    fun update(dt: Float) {
        position.x += (velocity.x * scale.x) * dt
        position.y += (velocity.y * scale.y) * dt
    }

    fun render(batch: Batch) {
        val x = ((position.x + cam.viewportWidth / 2 - cam.position.x) * scale.x) % bgTexture.regionWidth
        val y = ((position.y + cam.viewportHeight / 2 - cam.position.y) * scale.y) % bgTexture.regionHeight

        val colOffset = if (x > 0) -1 else 0
        val rowOffset = if (y > 0) -1 else 0

        for (row in 0..numDrawY) {
            for (col in 0..numDrawX) {
                batch.draw(bgTexture, x + (col + colOffset) * bgTexture.regionWidth,
                        y + (row + rowOffset) * bgTexture.regionHeight)
            }
        }
    }

    private fun fixBleeding(texture: TextureRegion) {
        val fix = 0.01f
        val invWidth = 1f / texture.texture.width
        val invHeight = 1f / texture.texture.height
        texture.setRegion((texture.regionX + fix) * invWidth,
                (texture.regionY + fix) * invHeight,
                (texture.regionX + texture.regionWidth - fix) * invWidth,
                (texture.regionY + texture.regionHeight - fix) * invHeight)
    }

}